README.md
# battery-indicator for Visual Studio Code

![screenshot](https://github.com/fBosch/vscode-battery-indicator/blob/master/screenshot.png?raw=true)

This package intended for use with the [Visual Studio Code](https://code.visualstudio.com/) editor and it displays battery status of your current device, which includes: power percentage, charging status and a visual indication bar.

TODO:
* Add settings for indicator symbols and color options

> Works for Linux, OS X and Windows.

CHANGELOG.md
# Change Log
All notable changes to the "batteryindicator" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Initial release

test/extension.test.ts
// The module 'assert' provides assertion methods from node
import * as assert from 'assert';
// You can import and use all API from the 'vscode' module
// as well as import your extension to test it
import * as vscode from 'vscode';

import { BatteryIndicator } from '../src/extension';

//
// Note: This example test is leveraging the Mocha test framework.
// Please refer to their documentation on https://mochajs.org/ for help.
//


// Defines a Mocha test suite to group tests of similar kind together
suite("Battery Indicator", () => {

    let batteryLevel = () => Promise.resolve(80)
    let chargingBattery = () => Promise.resolve(true)

    let noBattery = () => Promise.reject(0)
    let dischargingBattery = () => Promise.resolve(false)

    let batteryIndicator = new BatteryIndicator(batteryLevel, chargingBattery)
    let batteryIndicatorOnDeviceWithNoBattery = new BatteryIndicator(noBattery, dischargingBattery)


    test('should start polling upon initialization', () => {
        assert(batteryIndicator.interval)
    })

    test('should stop polling and clear interval instance when the current device has no battery', () => {
        assert(!batteryIndicatorOnDeviceWithNoBattery.interval)
    })

    test('should show charging icon when power cable is plugged ind', () => {
        assert(batteryIndicator.getVisualIndicator(100, true).includes(BatteryIndicator.CHARGING_SYMBOL))
        assert(batteryIndicator.getVisualIndicator(80, true).includes(BatteryIndicator.CHARGING_SYMBOL))
    })
   
   test('should represent percentage of battery (rounded up) as an ascii string', () => {
        assert(batteryIndicator.getVisualIndicator(100, false).includes(BatteryIndicator.BATTERY_SECTION_FULL_SYMBOL.repeat(10)))
        assert(batteryIndicator.getVisualIndicator(69, true).includes(BatteryIndicator.BATTERY_SECTION_FULL_SYMBOL.repeat(7)))
        assert(batteryIndicator.getVisualIndicator(10, true).includes(BatteryIndicator.BATTERY_SECTION_FULL_SYMBOL.repeat(1)))
        assert(batteryIndicator.getVisualIndicator(7, true).includes(BatteryIndicator.BATTERY_SECTION_FULL_SYMBOL.repeat(1)))
    })

    suite('Power Color', () => {
        test('should be dark green at 100% power', () => {
            assert.equal(batteryIndicator.getPowerColor(100), BatteryIndicator.COLORS.full)
        })
        test('should be light green at 80% to 99% power', () => {
            assert.equal(batteryIndicator.getPowerColor(80), BatteryIndicator.COLORS.high)
            assert.equal(batteryIndicator.getPowerColor(90), BatteryIndicator.COLORS.high)
        })
        test('should be yellow at 50% to 79% power ', () => {
            assert.equal(batteryIndicator.getPowerColor(50), BatteryIndicator.COLORS.medium)
            assert.equal(batteryIndicator.getPowerColor(79), BatteryIndicator.COLORS.medium)
        })
        test('should be red at 30% to 49% power', () => {
            assert.equal(batteryIndicator.getPowerColor(30), BatteryIndicator.COLORS.low)
            assert.equal(batteryIndicator.getPowerColor(49), BatteryIndicator.COLORS.low)
        })
        test('should be dark red below and equals 15% powe r', () => {
            assert.equal(batteryIndicator.getPowerColor(1), BatteryIndicator.COLORS.veryLow)
            assert.equal(batteryIndicator.getPowerColor(15), BatteryIndicator.COLORS.veryLow)
        })
    })
})

src/extension.ts
import * as execa from 'execa'
import * as isCharging from 'is-charging'
import * as linuxBattery from 'linux-battery'
import * as osxBattery from 'osx-battery'
import * as padEnd from 'pad-end'
import * as powerInfo from 'node-power-info'
import * as toDecimal from 'to-decimal'

import { Disposable, ExtensionContext, StatusBarAlignment, StatusBarItem, TextDocument, commands, window } from 'vscode'

export function activate(context: ExtensionContext) {
    console.log('🔋  battery-indicator is now active!')

    const unix = () => new Promise(resolve => powerInfo.getChargeStatus(batteryStats => resolve(batteryStats[0].powerLevel)))

    const win = () => execa.stdout('WMIC', ['Path', 'Win32_Battery', 'Get', 'EstimatedChargeRemaining']).then(stdout => {
        if (!stdout) {
            return Promise.reject(new Error('No battery could be found'));
        }
        stdout = parseFloat(stdout.trim().split('\n')[1]);
        return toDecimal(stdout > 100 ? 100 : stdout) * 100
    })

    let batteryLevelStateChecker
    if (process.platform === 'darwin' || process.platform === 'linux') {
        batteryLevelStateChecker = unix
    } else {
        batteryLevelStateChecker = win
    }

    let batteryIndicator = new BatteryIndicator(batteryLevelStateChecker, isCharging)
    let controller = new BatteryIndicatorController(batteryIndicator)

    context.subscriptions.push(controller)
    context.subscriptions.push(batteryIndicator)
}

export class BatteryIndicatorController {
    private _batteryIndicator: BatteryIndicator
    private _disposable: Disposable

    constructor(batteryIndicator: BatteryIndicator) {
        this._batteryIndicator = batteryIndicator
        this._batteryIndicator.updateStatus()

        let subscriptions: Disposable[] = []
        window.onDidChangeActiveTextEditor(this._onEvent, subscriptions)
        window.onDidChangeActiveTextEditor(this._onEvent, subscriptions)

        this._batteryIndicator.updateStatus()

        this._disposable = Disposable.from(...subscriptions)
    }

    dispose() {
        this._disposable.dispose()
    }

    private _onEvent() {
        this._batteryIndicator.updateStatus()
    }

}

export class BatteryIndicator {

    static POLLING_INTERVAL = 30000
    static BATTERY_SECTION_FULL_SYMBOL = '|'
    static BATTERY_SECTION_EMPTY_SYMBOL = ' '
    static BAR_LENGTH = 10
    static CHARGING_SYMBOL = '⚡'
    static COLORS = {
        full: '#8AE66C',
        high: '#D8FA63',
        medium: '#FED83A',
        low: '#FD9943',
        veryLow: '#FD5324'
    }

    private _battery
    private _statusBarItem: StatusBarItem
    private _pollingInterval
    private _batteryPercentage
    private _chargingState

    public interval
    public getBatteryState
    public getChargingState

    constructor(getBatteryState, getChargingState) {
        this.getBatteryState = getBatteryState
        this.getChargingState = getChargingState
        this.updateStatus()
        this.pollingInterval = BatteryIndicator.POLLING_INTERVAL
    }

    set pollingInterval(pollingInterval) {
        if (pollingInterval) {
            pollingInterval = Math.max(pollingInterval, 1)
            this._pollingInterval = pollingInterval
            this.startPolling()
        }
    }

    get pollingInterval() {
        return this._pollingInterval
    }

    stopPolling() {
        if (this.interval) {
            clearInterval(this.interval)
            this.interval = false
        }
    }

    startPolling() {
        if (this.interval) {
            clearInterval(this.interval)
        }
        this.interval = setInterval(() => this.updateStatus(), this.pollingInterval)
    }

    updateStatus() {
        if (!this._statusBarItem) {
            this._statusBarItem = window.createStatusBarItem(StatusBarAlignment.Right)
        }

        let batteryLevel = this.getBatteryState().then(level => this._batteryPercentage = level),
            chargingState = this.getChargingState().then(state => this._chargingState = state)

        Promise.all([batteryLevel, chargingState]).then(() => {
            this.updateStatusText(this._batteryPercentage, this._chargingState)
            this._statusBarItem.color = this.getPowerColor(this._batteryPercentage)
            this._statusBarItem.show()
        }, () => {
            console.warn('Current device does not have battery')
            this.dispose()
        })
    }

    getVisualIndicator(percentage, chargingState) {
        let bar = BatteryIndicator.BATTERY_SECTION_FULL_SYMBOL.repeat(BatteryIndicator.BAR_LENGTH)
        let currentLevel = padEnd(BatteryIndicator.BATTERY_SECTION_FULL_SYMBOL.repeat(Math.round((bar.length / 100) * percentage)), bar.length, BatteryIndicator.BATTERY_SECTION_EMPTY_SYMBOL)
        return `[${currentLevel}] ${chargingState ? BatteryIndicator.CHARGING_SYMBOL : ''}`
    }

    updateStatusText(percentage, chargingState) {
        if (percentage) {
            this._statusBarItem.text = `${percentage}% ${this.getVisualIndicator(percentage, chargingState)}`
        } else {
            console.warn(`battery-indicator: invalid value: ${percentage}`)
        }
    }

    getPowerColor(percentage) {
        if (percentage === 100) {
            return BatteryIndicator.COLORS.full
        } else if (percentage >= 80) {
            return BatteryIndicator.COLORS.high
        } else if (percentage >= 50) {
            return BatteryIndicator.COLORS.medium
        } else if (percentage >= 30) {
            return BatteryIndicator.COLORS.low
        } else if (percentage <= 15) {
            return BatteryIndicator.COLORS.veryLow
        }
        return null
    }

    dispose() {
        this.stopPolling()
        this._statusBarItem.dispose()
    }


}

export function deactivate() {
}